### PR TITLE
SONARJAVA-1498 Authorize exceptions when method is overriding

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ThrowCheckedExceptionCheck.java [hunk 1]
```diff
@@ -41,18 +46,53 @@
 @SqaleConstantRemediation("1h")
 public class ThrowCheckedExceptionCheck extends SubscriptionBaseVisitor {
 
+  private Stack<MethodTree> methods = new Stack<>();
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ThrowCheckedExceptionCheck.java]: MAJOR Replace the synchronized class "Stack" by an unsynchronized one such as "Deque". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1149)

File: java-checks/src/main/java/org/sonar/java/checks/ThrowCheckedExceptionCheck.java [hunk 2]
```diff
@@ -41,18 +46,53 @@
...
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD)) {
+      methods.pop();
+    }
+  }
+
+  private static boolean isFromMethodOverride(Type exceptionType, Stack<MethodTree> methods) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ThrowCheckedExceptionCheck.java]: MAJOR Replace the synchronized class "Stack" by an unsynchronized one such as "Deque". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1149)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/ThrowCheckedExceptionCheck.java [hunk 3]
```diff
@@ -41,18 +47,53 @@
...
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD)) {
+      methods.pop();
+    }
+  }
+
+  private static boolean isFromMethodOverride(Type exceptionType, Deque<MethodTree> methods) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/ThrowCheckedExceptionCheck.java]: If you don't make this static, no need for the second parameter as methods is a field of your class.

**Pull request merged**